clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -472,6 +472,12 @@ private static int parseAcks(String acksString) {
             // handling exceptions and record the errors;
             // for API exceptions return them in the future,
             // for other exceptions throw directly
+        } catch (TimeoutException e) {
+            this.errors.record();
+            this.metrics.sensor("buffer-exhausted-records").record();
```

Comment:
If I understand you correctly, that's indeed what I had in mind. We throw `BufferExhaustedException` from `BufferPool.allocate`. To make it compatible with code that expects a `TimeoutException`, we make it a subclass of `TimeoutException`. And then here we catch `BufferExhaustedException`.

That would mean not deprecating `BufferExhaustedException`, but updating its documentation to say that it's thrown when a buffer allocation times out.

What do you think @junrao?